internal/provider/data_source_gitlab_release_links.go

```diff
@@ -0,0 +1,82 @@
+package provider
+
+import (
+	"context"
+	"log"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/xanzy/go-gitlab"
+)
+
+var _ = registerDataSource("gitlab_release_links", func() *schema.Resource {
+
+	return &schema.Resource{
+		Description: `The ` + "`gitlab_release_links`" + ` data source allows get details of release links.
+
+**Upstream API**: [GitLab REST API docs](https://docs.gitlab.com/ee/api/releases/links.html)`,
+
+		ReadContext: dataSourceGitlabReleaseLinksRead,
+		Schema: map[string]*schema.Schema{
+			"project": {
+				Description: "The ID or [URL-encoded path of the project](https://docs.gitlab.com/ee/api/index.html#namespaced-path-encoding).",
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+			},
+			"tag_name": {
+				Description: "The tag associated with the Release.",
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+			},
+			"release_links": {
+				Description: "List of release links",
+				Type:        schema.TypeList,
+				Computed:    true,
+				Elem: &schema.Resource{
+					Schema: datasourceSchemaFromResourceSchema(gitlabReleaseLinkGetSchema(), nil, nil),
+				},
+			},
+		},
+	}
+})
+
+func dataSourceGitlabReleaseLinksRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*gitlab.Client)
+
+	project := d.Get("project").(string)
+	tagName := d.Get("tag_name").(string)
+	options := gitlab.ListReleaseLinksOptions(
+		gitlab.ListOptions{
+			PerPage: 20,
+			Page:    1,
+		})
+
+	var releaseLinks []*gitlab.ReleaseLink
+	for options.Page != 0 {
+		paginatedReleaseLinks, resp, err := client.ReleaseLinks.ListReleaseLinks(project, tagName, &options, gitlab.WithContext(ctx))
+		if err != nil {
+			return diag.FromErr(err)
+		}
```

Comment:
More a question for @timofurrer - Right now most of our "plural" data sources result in a 404 if they return 0 results due to how the underlying APIs work.. This can cause an issue in configurations where it's perfectly legit to have 0 results for a value. 

Is this really how we want these data sources to work, or should we catch 404 here using the `is404` function and return and break the loop which will just result in an empty list. The empty list will still cause an error in terraform configuration that _expect_ a value (because the downstream resources would break), but would still function perfectly fine in other scenarios where you may be just using the values as a loop in a `count` or `foreach` meta configuration.

Most of the other major cloud provider configurations are... inconsistent in this regard as well, with some data sources erroring, and some returning an empty array.

Reply:
Done. Additionally, I've added a test for this case.

Reply:
@PatrickRice-KSC @Moon1706 Yes, I agree to your judgement. My only concern where a `404` indeed is an error is if it is returned from any other page than the first, because that should actually be "accessible" - w.r.t. permissions and existence. 

But for the sake of simplicity I think it's "okay" to handle it as you've implemented it right now.

Reply:
@timofurrer If you want, I can append an additional section of condition (`if is404(err) && (options.Page == 1)`)

Reply:
@Moon1706 Yes, let's do that!

Reply:
Agreed, I like this pattern!

Reply:
Done